### PR TITLE
commands/block: use CIDv1 with custom mhtype

### PR DESCRIPTION
**Pull request by magik6k**
Closes https://github.com/ipfs/go-ipfs/issues/4741 ~Kubuxu

### CODE REVIEW

**Comment by Stebalien** [comment]
Shouldn't this actually just return an error if the user passes `-f v0 --mhtype=other-hash --mhlen=other-len`?

File: core/commands/block.go [hunk 1]
```diff
@@ -157,27 +159,32 @@ It reads from stdin, and <key> is a base58 encoded multihash.
...
+			err := fmt.Errorf("unrecognized multihash function: %s", mhtype)
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
 		var pref cid.Prefix
 		pref.Version = 1
 
-		format, _ := req.Options["format"].(string)
+		format, userFormat := req.Options["format"].(string)
```
Comment (Stebalien) [inline core/commands/block.go]: Unfortunately, `userFormat` will always be true there is a default value. The only way I know to fix this, is to simply remove the default.

File: test/sharness/t0050-block.sh [hunk 2]
```diff
@@ -209,4 +209,16 @@ test_expect_success "no panic in output" '
   test_expect_code 1 grep "panic" stat_out
 '
 
+test_expect_success "can set multihash type and length on block put without format" '
+  HASH=$(echo "foooo" | ipfs block put --mhtype=sha3 --mhlen=16)
```
Comment (Kubuxu) [inline test/sharness/t0050-block.sh]: 16 will be too short of a hash length after cid changes come through, please change it to 20.

Comment (magik6k) [reply test/sharness/t0050-block.sh]: done

File: core/commands/block.go [hunk 3]
```diff
@@ -121,14 +122,17 @@ var blockPutCmd = &cmds.Command{
...
+than 'sha2-256' or format to anything other than 'v0' will result in CIDv1.
 `,
 	},
 
 	Arguments: []cmdkit.Argument{
 		cmdkit.FileArg("data", true, false, "The data to be stored as an IPFS block.").EnableStdin(),
 	},
 	Options: []cmdkit.Option{
-		cmdkit.StringOption("format", "f", "cid format for blocks to be created with.").WithDefault("v0"),
+		cmdkit.StringOption("format", "f", "cid format for blocks to be created with.").WithDefault(""),
```
Comment (Stebalien) [inline core/commands/block.go]: I wouldn't specify a default at all. As it is, passing `-f ""` will be considered valid even though it makes no sense.

If you don't specify a default, you can tell if the parameter was passed with `format, formatSet := req.Options["format"].(string)`.

**Comment by schomatis** [comment]
This PR looks correct to me (and it fixes the addressed issue) but let me raise an off-topic question: shouldn't part of this prefix preprocessing logic be in [`go-cid`](https://github.com/ipfs/go-cid#creating-a-cid-from-scratch)?

It seems to me there are a couple of hard-coded pieces of information of what constitutes a valid CID prefix that the `ipfs block` command doesn't need to know because it shouldn't be its responsibility to check for it, e.g., the (now) default CIDv1 specified here (`pref.Version = 1`) will have to be updated if (or when) a new CID version is released.

The `ipfs block` command doesn't do anything with the version and hash function except forward them to `Prefix`, which is the one that computes the hash/CID that `ipfs block` is really interested in (to create block).

**Comment by Kubuxu** [comment]
I think most of the processing here is due to 1. us wanting to provide good error messages for user 2. be lenient with the user input. One check in question (CIDv0 can only be full length SHA256) is currently not implemented but we are in process of changing that (https://github.com/ipfs/go-cid/pull/40/files#diff-beb384810133d725af6b502e9b6e3de8R139)

I am not sure if go-cid would be right place to cover all of those areas but having a helper function somewhere in go-ipfs would not hurt.

**Comment by schomatis** [comment]
> us wanting to provide good error messages for user

That's a good point I hadn't considered, but looking at the errors:

* `unrecognized multihash function`
* `unrecognized format`
* `cannot generate CIDv0 with non-sha256 hash function`

I think all of them can be outsourced without losing expressiveness for the user (the `missing option "mhlen"` seems to me the only error that truly belongs here).


> I am not sure if go-cid would be right place to cover all of those areas but having a helper function somewhere in go-ipfs would not hurt.

That's fine, my main point wasn't where this logic should be placed but that it should be encapsulated somewhere outside this command function.

Anyway, maybe this should be discussed in another issue after this PR is merged, trying to change all this here in this PR now would be to messy, but I wanted to leave this suggestion documented somewhere.

**Comment by Stebalien** [comment]
I think the issue here is that we want to catch all of these errors early *and* provide reasonable defaults (that may change based on the command in question). However, we probably could have something in `go-cid` that takes in a set of CID "constraints" and spits out either an error or the set of CID parameters.

An alternative solution would be to add flag "sets" to go-ipfs-cmds and create a flag set for CID options somewhere.

**Comment by whyrusleeping** [comment]
First PR merged of the 0.4.15 cycle goes to @magik6k :) 👏

**Pull request merged**